dask/array/random.py

```diff
@@ -360,124 +675,398 @@ def permutation(self, x):
+    state_data = bit_generator.state
+    bit_generator = type(bit_generator)()
+    bit_generator.state = state_data
+    state = _rng_from_bitgen(bit_generator)
+    return state.shuffle(x, axis=axis)
+
+
+def _spawn_bitgens(bitgen, n_bitgens):
+    seeds = bitgen._seed_seq.spawn(n_bitgens)
+    bitgens = [type(bitgen)(seed) for seed in seeds]
+    return bitgens
+
+
+def _apply_random_func(rng, funcname, bitgen, size, args, kwargs):
+    """Apply random module method with seed"""
+    if isinstance(bitgen, np.random.SeedSequence):
+        bitgen = rng(bitgen)
+    rng = _rng_from_bitgen(bitgen)
+    func = getattr(rng, funcname)
+    return func(*args, size=size, **kwargs)
 
 
 def _apply_random(RandomState, funcname, state_data, size, args, kwargs):
     """Apply RandomState method with seed"""
     if RandomState is None:
-        RandomState = np.random.RandomState
+        RandomState = array_creation_dispatch.RandomState
     state = RandomState(state_data)
     func = getattr(state, funcname)
     return func(*args, size=size, **kwargs)
 
 
-_cached_random_states = {}
+def _choice_rng(state_data, a, size, replace, p, axis, shuffle):
+    state = _rng_from_bitgen(state_data)
+    return state.choice(a, size=size, replace=replace, p=p, axis=axis, shuffle=shuffle)
+
 
+def _choice_rs(state_data, a, size, replace, p):
+    state = array_creation_dispatch.RandomState(state_data)
+    return state.choice(a, size=size, replace=replace, p=p)
+
+
+def _choice_validate_params(state, a, size, replace, p, axis, chunks):
+    dependencies = []
+    # Normalize and validate `a`
+    if isinstance(a, Integral):
+        if isinstance(state, Generator):
+            if state._backend_name == "cupy":
+                raise NotImplementedError(
+                    "`choice` not supported for cupy-backed `Generator`."
+                )
+            meta = state._backend.random.default_rng().choice(1, size=(), p=None)
+        elif isinstance(state, RandomState):
+            # On windows the output dtype differs if p is provided or
+            # # absent, see https://github.com/numpy/numpy/issues/9867
+            dummy_p = state._backend.array([1]) if p is not None else p
+            meta = state._backend.random.RandomState().choice(1, size=(), p=dummy_p)
+        else:
+            raise ValueError("Unknown generator class")
+        len_a = a
+        if a < 0:
+            raise ValueError("a must be greater than 0")
+    else:
+        a = asarray(a)
+        a = a.rechunk(a.shape)
+        meta = a._meta
+        if a.ndim != 1:
+            raise ValueError("a must be one dimensional")
+        len_a = len(a)
+        dependencies.append(a)
+        a = a.__dask_keys__()[0]
+
+    # Normalize and validate `p`
+    if p is not None:
+        if not isinstance(p, Array):
+            # If p is not a dask array, first check the sum is close
+            # to 1 before converting.
+            p = np.asarray(p)
```

Comment:
Will `p` ever be a CuPy array? If so this will break, and we will need to use [`asarray_safe` with a `like=` reference array](https://github.com/dask/dask/blob/5a6bbdc1f2366f0977b38d6bb95a343b6e5c5454/dask/array/utils.py#L457-L468) instead.

Reply:
It is a good idea to future proof in any case. see c39dafe